src/Symfony/Component/Cache/Adapter/RedisTagAwareAdapter.php

```diff
@@ -161,46 +156,33 @@ protected function doDelete(array $ids, array $tagData = []): bool
      */
     protected function doInvalidate(array $tagIds): bool
     {
-        if (!$this->redisServerSupportSPOP()) {
-            return false;
-        }
-
-        // Pop all tag info at once to avoid race conditions
-        $tagIdSets = $this->pipeline(static function () use ($tagIds) {
-            foreach ($tagIds as $tagId) {
-                // Client: Predis or PHP Redis 3.1.3+ (https://github.com/phpredis/phpredis/commit/d2e203a6)
-                // Server: Redis 3.2 or higher (https://redis.io/commands/spop)
-                yield 'sPop' => [$tagId, self::POP_MAX_LIMIT];
+        // Rename all sets so we can safely avoid race conditions
+        $this->pipeline(static function () use ($tagIds) {
+            $uniqueToken = bin2hex(random_bytes(10));
+            foreach ($tagIds as $key => $tagId) {
+                $tagIds[$key] = $temporaryTagId = $tagId . ':' . $uniqueToken;
+                yield 'rename' => [$tagId, $temporaryTagId];
```

Comment:
Correct me if I'm off base here but you'd better use [sMove](https://github.com/phpredis/phpredis/blob/d2e203a630437f4de657c9b0b34d689970a56b21/README.markdown#smove) here to rename this in batch

Reply:
I'll have a look

Reply:
sMove seems on closer look irrelevant. With the rename I move the whole Set in one operation to be able to afterwards iteratively read it without having to worry about race conditions of other processes adding keys to the tag set. In the end I delete the temporary set.